text/0000-safer-transmute.md

```diff
@@ -0,0 +1,1796 @@
+    }
+};
+```
+Given these two stability promises, we are free to modify the alignment of `Foo` in SemVer-minor changes without running any risk of breaking dependent crates.
+
+
+## Mechanisms of Transmutation
+
+Two traits provide mechanisms for transmutation between types: 
+```rust
+// this trait is implemented automagically by the compiler
+#[lang = "transmute_from"]
+pub unsafe trait TransmuteFrom<Src: ?Sized, Neglect = ()>
+where
+    Neglect: TransmuteOptions,
+{
+    #[inline(always)]
+    fn transmute_from(src: Src) -> Self
+    where
+        Src: Sized,
+        Self: Sized,
+        Neglect: SafeTransmuteOptions,
+    {
+        unsafe { Self::unsafe_transmute_from(src) }
+    }
+
+    #[inline(always)]
+    unsafe fn unsafe_transmute_from(src: Src) -> Self
+    where
+        Src: Sized,
+        Self: Sized,
+        Neglect: TransmuteOptions,
+    {
+        use core::mem::ManuallyDrop;
+
+        #[repr(C)]
+        union Transmute<Src, Dst> {
+            src: ManuallyDrop<Src>,
+            dst: ManuallyDrop<Dst>,
+        }
+
+        unsafe {
+            ManuallyDrop::into_inner(Transmute { src: ManuallyDrop::new(src) }.dst)
+        }
+    }
+}
+
+// implemented in terms of `TransmuteFrom`
+pub unsafe trait TransmuteInto<Dst: ?Sized, Neglect = ()>
+where
+    Neglect: TransmuteOptions,
+{
+    fn transmute_into(self) -> Dst
+    where
+        Self: Sized,
+        Dst: Sized,
+        Neglect: SafeTransmuteOptions;
+
+    unsafe fn unsafe_transmute_into(self) -> Dst
+    where
+        Self: Sized,
+        Dst: Sized,
+        Neglect: TransmuteOptions;
+}
+
+unsafe impl<Src, Dst, Neglect> TransmuteInto<Dst, Neglect> for Src
+where
+    Src: ?Sized,
+    Dst: ?Sized + TransmuteFrom<Src, Neglect>,
+    Neglect: TransmuteOptions,
+{
+    ...
+}
+```
+
+In the above definitions, `Src` represents the source type of the transmutation, `Dst` represents the destination type of the transmutation, and `Neglect` is a parameter that [encodes][options] which static checks the compiler ought to neglect when considering if a transmutation is valid. The default value of `Neglect` is `()`, which reflects that, by default, the compiler does not neglect *any* static checks.
+
+### Neglecting Static Checks
+[options]: #Neglecting-Static-Checks
+
+The default value of the `Neglect` parameter, `()`, statically forbids transmutes that are unsafe, unsound, or unstable. However, you may explicitly opt-out of some static checks:
+
+| Transmute Option    | Compromises | Usable With                                             |
+|---------------------|-------------|---------------------------------------------------------|
+| `NeglectStabilty`   | Stability   | `transmute_{from,into}`, `unsafe_transmute_{from,into}` |
+| `NeglectAlignment`  | Safety      | `unsafe_transmute_{from,into}`                          |
+| `NeglectValidity`   | Soundness   | `unsafe_transmute_{from,into}`                          |
```

Comment:
I'm wondering if there's a missing option here. As far as I can tell "Validity" refers specifically to bit validity - i.e., is the in-memory representation a valid representation of this type. But there's also higher-level validity concerns such as correctness. A `NonZeroUsize` is a valid `&NonZeroU8`, it has the correct alignment proprieties, and it is stable, but of course it might not point to a valid `NonZeroU8`. Do we need to have separate neglect options for higher-level invariants?

Reply:
>  A NonZeroUsize is a valid &NonZeroU8, it has the correct alignment proprieties, and it is stable, but of course it might not point to a valid NonZeroU8.

That is not decided yet -- it might be that `&NonZeroU8` actually must point to something non-zero immediately when constructed, making this part of the fundamental validity rules. In fact that is the status quo, as relaxing this in the future is easier than strengthening it. See https://github.com/rust-lang/unsafe-code-guidelines/issues/77.